### DISCUSSION TITLE
About part of UI component license

### DISCUSSION THREAD

**Discussion by huacnlee** [zed-industries/zed#13694] [Show and tell]
Hi Zed team,

I've worked on to create a common UI component based on GPUI in recent days.

Ref #9877 

https://github.com/huacnlee/gpui-component

Because our company has observed the GPUI foundation long time ago, until I have fortunately to contribute some codes into the Zed project, and to leaned more about the GPUI architecture.

And we are keep looking for a good UI framework lie iced, WinUI, Flutter, QT, etc. to build our commercial project. We have done a lot of research and test on those UI frameworks, but we still can't find a good one to meet our requirement by some reasons.

The GPUI is a great project, the design and performance is so great, and the Zed is also too. That give me a lot of inspiration organizing the large scale project.

So, it gives us a confidence to consider use the GPUI in our commercial project to build a large cross-platform desktop application.

----

Currently, I have already done the most useful components of the work. It is looks like ready to build a complex UI.

I have read a lot of the Zed UI code, and some parts are split from the Zed UI.

And those parts are GPL licensed:

- crates/workspace (include the workspace, dock, pane, item, etc.)
- scrollbar
- picker
- title_bar

Is there any possibility to change the license to Apache-2.0 in that I split part?

This is important for us to use the GPUI in our project. And some part have a lot of code, it we rebuild it will block on it a long time.

I am truly hoped to get your permission. 

And we will continue to contribute to the GPUI community to share our work on the base UI component.

**Comment by huacnlee** [comment]
Here is some demo:

<https://github.com/huacnlee/gpui-app/assets/5518/0273e031-4426-4ab5-a41c-f7cbbb0e55bc>

<https://github.com/huacnlee/gpui-component/assets/5518/51622a5e-f51d-4ede-8cae-04cae703f8aa>

**Reply by huacnlee** [reply]
These components are relatively common basic UI components, which are basically needed by many desktop applications. Currently, GPUI has built-in relatively basic functions, which makes it difficult for other people with ideas to quickly use it to build projects.

> For example, Input, if it weren't for the help of @ConradIrwin in #13534 last week, I might have to study it for a long time.

I think if it has a more broad License, it will also be beneficial for the GPUI project to be used by more and more projects in the community to build great cross-platform desktop projects.

**Comment by maxdeviant** [comment]
We have no plans to change the licensing of `ui` or any of the other mentioned crates.

These are all built to be Zed-specific and aren't intended for use outside of Zed. The choice of license reflects this.

At some point we want to build a more general-purpose UI library (more along the lines of [Radix](https://www.radix-ui.com/) for React) that would be open-sourced with the intent for it to be used in many GPUI projects under a more permissive license, but right now we're primarily focused on Zed itself.

**Reply by huacnlee** [reply]
As you said, GPUI also plans to provide more general GPUI basic UI components.

I don't mean to change the `ui` crate and other libraries in the Zed project.

I just hope you can consider allowing several components I have extracted to use Apache License.

In this project:

https://github.com/huacnlee/gpui-component

- dock
- scrollbar
- picker
- title_bar

**Reply by huacnlee** [reply]
In fact, I have seen that the Zed team has plans to do something similar.

If this is consistent with your purpose, perhaps this project can show the community and provide basic UI components so that GPUI can start to be used in more projects.

At least the Zed team currently needs focus on the Zed part and does not have much time to promote GPUI basic components. These components can allow people who want to use GPUI to start using it.

**Reply by maxdeviant** [reply]
> I just hope you can consider allowing several components I have extracted to use Apache License.
> 
> In this project:
> 
> https://github.com/huacnlee/gpui-component
> 
> * dock
> * scrollbar
> * picker
> * title_bar

Taking the contents of the crates wholesale and relicensing them is a violation of the license, and I don't think we can legally permit it.

**Reply by huacnlee** [reply]
Ok, I get it.

**Comment by xtrasmal** [comment]
Somebody should tell you that you are awesome. @huacnlee

**Comment by auriium2** [comment]
Ditto, coming back to this and it's increasingly obvious that some of the contributions of gpui-component should be in the zed ui core, since they would be incredibly useful for shipping some of the newer zed features (latex support, webviews, etc). This should be reconsidered.
